uaclient/system.py

```diff
@@ -370,12 +371,22 @@ def write_file(filename: str, content: str, mode: int = 0o644) -> None:
     @param content: The content to write to the file.
     @param mode: The filesystem mode to set on the file.
     """
-    logging.debug("Writing file: %s", filename)
-    os.makedirs(os.path.dirname(filename), exist_ok=True)
-    with open(filename, "wb") as fh:
-        fh.write(content.encode("utf-8"))
-        fh.flush()
-    os.chmod(filename, mode)
+    tmpf = None
+    try:
+        tmpf = tempfile.NamedTemporaryFile(mode="wb", delete=False)
+        logging.debug(
+            "Writing file %s atomically via tempfile %s", filename, tmpf.name
+        )
+        tmpf.write(content.encode("utf-8"))
+        tmpf.flush()
+        tmpf.close()
+        os.chmod(tmpf.name, mode)
+        os.makedirs(os.path.dirname(filename), exist_ok=True)
+        os.rename(tmpf.name, filename)
+    except Exception as e:
```

Comment:
Maybe this is an overkill, but should we add an unit test for this code ? Just for the case where the write part fails and we call the `flush` method.

But let me know what you think

Reply:
Good idea, I'll add one

Reply:
Yes, I think that covers it perfectly